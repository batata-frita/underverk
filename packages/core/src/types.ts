export interface Reference {
  type: 'reference'
  value: string
}

export interface Operation {
  type: 'operation'
  name: string
  arguments: Expression[]
}

export type Expression = Operation | Reference

export interface Argument {
  name: string
}

export interface Declaration {
  name: string
  value: Expression
}

export interface Literal {
  name: string
  value: string | number | boolean | object
}

export interface Function {
  name: string
  arguments: Argument[]
  composition: Operation[]
}

export interface StaticNodeChild {
  type: 'staticNode'
  element: string
  props: Declaration[]
  children: Child[]
}

export interface DynamicNodeChild {
  type: 'dynamicNode'
  element: string
  dependency: string
}

export type Child = StaticNodeChild | DynamicNodeChild | Expression

export interface UpdateFunction {
  name: string

  // this points to a function
  transformation: string
}

export interface State {
  updateFunctions: UpdateFunction[]
  defaultValue: Expression
}

export interface Effect {
  type: 'log'
  dependency: string
  handler: string
}

export interface Computed {
  name: string
  operation: string
  arguments: Reference[]
}

export interface Component {
  name: string
  props: Argument[]

  literals: Literal[]
  functions: Function[]
  state?: State
  computed: Computed[]

  effects: Effect[]
  children: Child[]
}

export interface Project {
  name: string
  components: Component[]
}
