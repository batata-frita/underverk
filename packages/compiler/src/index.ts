import * as t from '@babel/types'
import template from '@babel/template'
import babelGenerate from '@babel/generator'
import {
  Component,
  State,
  Expression,
  Declaration,
  Effect,
  Child,
  Literal,
  Computed,
  Reference,
  Function,
  UpdateFunction,
  Argument,
} from './types'
export * from './types'
export * from './manipulations'

export const generate = (ast: t.VariableDeclaration): string => babelGenerate(ast).code

export const compile = (componentAst: Component): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(componentAst.name),
      t.arrowFunctionExpression(
        [compileProps(componentAst.props)],
        t.blockStatement([
          ...componentAst.literals.map(compileLiteral),
          ...componentAst.functions.map(compileFunction(componentAst.props)),
          ...(componentAst.state ? compileState(componentAst.state) : []),
          ...componentAst.computed.map(compileComputed),
          ...componentAst.effects.map(compileEffect),
          compileChildren(componentAst.children),
        ]),
      ),
    ),
  ])

const compileProps = (propsAst: Argument[]): t.ObjectPattern =>
  t.objectPattern(propsAst.map(prop => t.objectProperty(t.identifier(prop.name), t.identifier(prop.name))))

const compileLiteral = (literalAst: Literal): t.VariableDeclaration => {
  switch (typeof literalAst.value) {
    case 'string':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.stringLiteral(literalAst.value)),
      ])

    case 'boolean':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.booleanLiteral(literalAst.value)),
      ])

    case 'number':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.numericLiteral(literalAst.value)),
      ])

    case 'object':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), template.expression(JSON.stringify(literalAst.value))()),
      ])

    case 'undefined':
      return t.variableDeclaration('let', [t.variableDeclarator(t.identifier(literalAst.name))])

    default:
      console.warn('ðŸ™ˆ literal captured by default case', literalAst)
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.stringLiteral(`${literalAst.value}`)),
      ])
  }
}

const compileFunction = (props: Argument[]) => (functionAst: Function): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(functionAst.name),
      t.callExpression(t.identifier('useCallback'), [
        t.arrowFunctionExpression(
          functionAst.arguments.map(argument => t.identifier(argument.name)),
          compileComposition(functionAst.composition),
        ),
        t.arrayExpression(props.map(({ name }) => t.identifier(name))),
      ]),
    ),
  ])

const compileComposition = (expressionAst: Expression[]): t.CallExpression =>
  t.callExpression(t.callExpression(t.identifier('compose'), expressionAst.map(compileExpression)), [
    t.identifier('event'),
  ])

const compileChildren = (childrenAst: Child[]): t.ReturnStatement =>
  t.returnStatement(t.jsxFragment(t.jsxOpeningFragment(), t.jsxClosingFragment(), childrenAst.map(compileChild)))

const compileChild = (childAst: Child): t.JSXElement | t.JSXText | t.JSXExpressionContainer => {
  switch (childAst.type) {
    case 'staticNode':
      const selfClosingTag = childAst.children.length === 0

      return t.jsxElement(
        t.jsxOpeningElement(t.jsxIdentifier(childAst.element), childAst.props.map(compileJsxProp), selfClosingTag),
        selfClosingTag ? undefined : t.jsxClosingElement(t.jsxIdentifier(childAst.element)),
        childAst.children.map(compileChild),
        selfClosingTag,
      )

    case 'dynamicNode':
      return t.jsxExpressionContainer(
        t.callExpression(t.identifier('dynamicNode'), [
          t.identifier(childAst.element),
          t.identifier(childAst.dependency),
        ]),
      )

    default:
      return t.jsxExpressionContainer(compileExpression(childAst))
  }
}

const compileJsxProp = (propAst: Declaration): t.JSXAttribute =>
  t.jsxAttribute(t.jsxIdentifier(propAst.name), t.jsxExpressionContainer(compileExpression(propAst.value)))

const compileEffect = (effectAst: Effect): t.ExpressionStatement =>
  t.expressionStatement(
    t.callExpression(t.identifier('useEffect'), [
      t.arrowFunctionExpression(
        [],
        t.blockStatement([
          t.expressionStatement(
            t.callExpression(t.identifier(effectAst.type), [
              t.identifier(effectAst.dependency),
              t.identifier(effectAst.handler),
            ]),
          ),
        ]),
      ),
      t.arrayExpression([t.identifier(effectAst.dependency)]),
    ]),
  )

const compileComputed = (computedAst: Computed): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(computedAst.name),
      t.callExpression(t.identifier(computedAst.operation), computedAst.arguments.map(compileReference)),
    ),
  ])

const compileExpression = (expressionAst: Expression): t.Expression => {
  switch (expressionAst.type) {
    case 'operation':
      return t.callExpression(t.identifier(expressionAst.name), expressionAst.arguments.map(compileExpression))

    case 'reference':
      return compileReference(expressionAst)

    default:
      console.warn('ðŸ™ˆ expression captured by default case', expressionAst)
      return t.stringLiteral('')
  }
}

const compileReference = (referenceAst: Reference): t.Identifier => t.identifier(referenceAst.value)

const compileState = (stateAst: State): t.VariableDeclaration[] => [
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.arrayPattern([t.identifier('state'), t.identifier('setState')]),
      t.callExpression(t.identifier('useState'), [compileExpression(stateAst.defaultValue)]),
    ),
  ]),
  ...stateAst.updateFunctions.map(compileUpdateFunction),
]

const compileUpdateFunction = (updateFunctionAst: UpdateFunction): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(updateFunctionAst.name),
      t.callExpression(t.identifier('useCallback'), [
        t.arrowFunctionExpression(
          [t.identifier('event')],
          t.callExpression(t.identifier('setState'), [
            t.objectExpression([
              t.spreadElement(t.identifier('state')),
              t.spreadElement(
                t.callExpression(t.identifier(updateFunctionAst.transformation), [
                  t.identifier('event'),
                  t.identifier('state'),
                ]),
              ),
            ]),
          ]),
        ),
        t.arrayExpression([t.identifier('state'), t.identifier(updateFunctionAst.transformation)]),
      ]),
    ),
  ])
