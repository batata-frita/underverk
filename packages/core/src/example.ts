import { Project, Component } from './types'

const Button: Component = {
  name: 'Button',
  props: [{ name: 'onClick' }],
  declarations: [
    {
      name: 'label',
      value: {
        type: 'operation',
        name: 'concat',
        arguments: [
          { type: 'literal', value: 'Hello' },
          { type: 'literal', value: 'World' },
        ],
      },
    },
  ],
  children: [
    {
      type: 'node',
      element: { type: 'literal', value: 'div' },
      props: [],
      children: [{ type: 'reference', value: 'label' }],
    },
  ],
}

export const project: Project = {
  name: 'My components',
  components: [Button],
}
