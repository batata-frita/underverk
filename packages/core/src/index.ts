import * as t from '@babel/types'
import { Component, State, Expression, Declaration } from './types'

export default (componentAst: Component): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(componentAst.name),
      t.arrowFunctionExpression(
        [],
        t.blockStatement([
          ...componentAst.states.map(compileState),
          ...componentAst.declarations.map(compileDeclaration),
        ]),
      ),
    ),
  ])

const compileDeclaration = (declarationAst: Declaration): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(t.identifier(declarationAst.name), compileExpression(declarationAst.value)),
  ])

const compileExpression = (expressionAst: Expression): t.Expression => {
  switch (expressionAst.type) {
    case 'literal':
      return t.stringLiteral(`${expressionAst.value}`)

    case 'function':
      return t.callExpression(t.identifier('callback'), [
        t.arrowFunctionExpression(
          expressionAst.arguments.map(argument => t.identifier(argument.name)),
          compileExpression(expressionAst.body),
        ),
      ])

    case 'operation':
      return t.callExpression(t.identifier(expressionAst.name), expressionAst.arguments.map(compileExpression))

    case 'reference':
      return t.identifier(expressionAst.value)

    default:
      console.warn('ðŸ™ˆ expression captured by default case', expressionAst)
      return t.stringLiteral('')
  }
}

const compileState = (stateAst: State): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.arrayPattern([t.identifier(stateAst.name), t.identifier(stateAst.updateFunction)]),
      t.callExpression(t.identifier('state'), [compileExpression(stateAst.defaultValue)]),
    ),
  ])
