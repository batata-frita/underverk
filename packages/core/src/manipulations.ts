import { Component } from './types'

export const getComponentSources = (componentAts: Component) => {
  return componentAts
}
