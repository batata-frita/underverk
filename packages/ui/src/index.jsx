import React from 'react'
import ReactDOM from 'react-dom'
import './index.css'
import App from './App'
/* eslint-disable import/no-webpack-loader-syntax */
import testProgram from '-!@underverk/loader!./testProgram.underverk.js'

console.log('ðŸ§±', testProgram)

ReactDOM.render(<App />, document.getElementById('root'))
