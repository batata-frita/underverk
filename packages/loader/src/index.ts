import { generateProgram } from '@underverk/compiler'

export default function(jsonSource: string) {
  return generateProgram(JSON.parse(jsonSource))
}
