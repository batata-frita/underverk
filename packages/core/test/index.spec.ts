import { compile, generate } from '../src'
import { App } from './examples/todoExample'
import { Button } from './examples/staticExample'

describe('statefulExample', () => {
  it('compiles', () => {
    expect(compile(App)).toMatchSnapshot()
  })

  it('generates', () => {
    expect(generate(compile(App))).toMatchSnapshot()
  })
})

describe('staticExample', () => {
  it('compiles', () => {
    expect(compile(Button)).toMatchSnapshot()
  })

  it('generates', () => {
    expect(generate(compile(Button))).toMatchSnapshot()
  })
})
