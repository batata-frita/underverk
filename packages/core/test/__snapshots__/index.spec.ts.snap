// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statefulExample compiles 1`] = `
Object {
  "declarations": Array [
    Object {
      "id": Object {
        "name": "App",
        "type": "Identifier",
      },
      "init": Object {
        "async": false,
        "body": Object {
          "body": Array [
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "valueKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "value",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "targetKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "target",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "listKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "list",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "currentKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "current",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "doneKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "done",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "falseKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "false",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "defaultCurrent",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "textKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "text",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "defaultState",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "[object Object]",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "submitLabel",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "Submit",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "oneKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "NumericLiteral",
                    "value": 1,
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "zeroKey",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "NumericLiteral",
                    "value": 0,
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "getValueFromEvent",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "async": false,
                    "body": Object {
                      "arguments": Array [
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "currentKey",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "objectOf",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "valueKey",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "get",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "targetKey",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "get",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "oneKey",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "nth",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                      ],
                      "callee": Object {
                        "name": "compose",
                        "type": "Identifier",
                      },
                      "type": "CallExpression",
                    },
                    "params": Array [
                      Object {
                        "name": "state",
                        "type": "Identifier",
                      },
                      Object {
                        "name": "event",
                        "type": "Identifier",
                      },
                    ],
                    "type": "ArrowFunctionExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "addTodo",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "async": false,
                    "body": Object {
                      "arguments": Array [
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "listKey",
                              "type": "Identifier",
                            },
                            Object {
                              "arguments": Array [
                                Object {
                                  "arguments": Array [
                                    Object {
                                      "name": "doneKey",
                                      "type": "Identifier",
                                    },
                                    Object {
                                      "name": "falseKey",
                                      "type": "Identifier",
                                    },
                                    Object {
                                      "arguments": Array [
                                        Object {
                                          "name": "textKey",
                                          "type": "Identifier",
                                        },
                                        Object {
                                          "arguments": Array [
                                            Object {
                                              "name": "currentKey",
                                              "type": "Identifier",
                                            },
                                            Object {
                                              "name": "state",
                                              "type": "Identifier",
                                            },
                                          ],
                                          "callee": Object {
                                            "name": "get",
                                            "type": "Identifier",
                                          },
                                          "type": "CallExpression",
                                        },
                                      ],
                                      "callee": Object {
                                        "name": "objectOf",
                                        "type": "Identifier",
                                      },
                                      "type": "CallExpression",
                                    },
                                  ],
                                  "callee": Object {
                                    "name": "set",
                                    "type": "Identifier",
                                  },
                                  "type": "CallExpression",
                                },
                                Object {
                                  "arguments": Array [
                                    Object {
                                      "name": "listKey",
                                      "type": "Identifier",
                                    },
                                    Object {
                                      "name": "state",
                                      "type": "Identifier",
                                    },
                                  ],
                                  "callee": Object {
                                    "name": "get",
                                    "type": "Identifier",
                                  },
                                  "type": "CallExpression",
                                },
                              ],
                              "callee": Object {
                                "name": "append",
                                "type": "Identifier",
                              },
                              "type": "CallExpression",
                            },
                          ],
                          "callee": Object {
                            "name": "set",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "currentKey",
                              "type": "Identifier",
                            },
                            Object {
                              "name": "defaultCurrent",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "set",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        Object {
                          "arguments": Array [
                            Object {
                              "name": "state",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "constant",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                      ],
                      "callee": Object {
                        "name": "compose",
                        "type": "Identifier",
                      },
                      "type": "CallExpression",
                    },
                    "params": Array [
                      Object {
                        "name": "event",
                        "type": "Identifier",
                      },
                      Object {
                        "name": "state",
                        "type": "Identifier",
                      },
                    ],
                    "type": "ArrowFunctionExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "elements": Array [
                      Object {
                        "name": "state",
                        "type": "Identifier",
                      },
                      Object {
                        "name": "updateState",
                        "type": "Identifier",
                      },
                    ],
                    "type": "ArrayPattern",
                  },
                  "init": Object {
                    "arguments": Array [
                      Object {
                        "name": "defaultState",
                        "type": "Identifier",
                      },
                    ],
                    "callee": Object {
                      "name": "useState",
                      "type": "Identifier",
                    },
                    "type": "CallExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "handleCurrentChange",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "arguments": Array [
                      Object {
                        "async": false,
                        "body": Object {
                          "arguments": Array [
                            Object {
                              "arguments": Array [
                                Object {
                                  "name": "event",
                                  "type": "Identifier",
                                },
                                Object {
                                  "name": "state",
                                  "type": "Identifier",
                                },
                              ],
                              "callee": Object {
                                "name": "getValueFromEvent",
                                "type": "Identifier",
                              },
                              "type": "CallExpression",
                            },
                          ],
                          "callee": Object {
                            "name": "setState",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        "params": Array [
                          Object {
                            "name": "event",
                            "type": "Identifier",
                          },
                        ],
                        "type": "ArrowFunctionExpression",
                      },
                      Object {
                        "elements": Array [
                          Object {
                            "name": "state",
                            "type": "Identifier",
                          },
                        ],
                        "type": "ArrayExpression",
                      },
                    ],
                    "callee": Object {
                      "name": "useCallback",
                      "type": "Identifier",
                    },
                    "type": "CallExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "updateClick",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "arguments": Array [
                      Object {
                        "async": false,
                        "body": Object {
                          "arguments": Array [
                            Object {
                              "arguments": Array [
                                Object {
                                  "name": "event",
                                  "type": "Identifier",
                                },
                                Object {
                                  "name": "state",
                                  "type": "Identifier",
                                },
                              ],
                              "callee": Object {
                                "name": "addToDo",
                                "type": "Identifier",
                              },
                              "type": "CallExpression",
                            },
                          ],
                          "callee": Object {
                            "name": "setState",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        "params": Array [
                          Object {
                            "name": "event",
                            "type": "Identifier",
                          },
                        ],
                        "type": "ArrowFunctionExpression",
                      },
                      Object {
                        "elements": Array [
                          Object {
                            "name": "state",
                            "type": "Identifier",
                          },
                        ],
                        "type": "ArrayExpression",
                      },
                    ],
                    "callee": Object {
                      "name": "useCallback",
                      "type": "Identifier",
                    },
                    "type": "CallExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "list",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "arguments": Array [
                      Object {
                        "name": "listKey",
                        "type": "Identifier",
                      },
                      Object {
                        "name": "state",
                        "type": "Identifier",
                      },
                    ],
                    "callee": Object {
                      "name": "get",
                      "type": "Identifier",
                    },
                    "type": "CallExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "expression": Object {
                "arguments": Array [
                  Object {
                    "async": false,
                    "body": Object {
                      "body": Array [
                        Object {
                          "expression": Object {
                            "arguments": Array [
                              Object {
                                "name": "list",
                                "type": "Identifier",
                              },
                              Object {
                                "name": "noop",
                                "type": "Identifier",
                              },
                            ],
                            "callee": Object {
                              "name": "log",
                              "type": "Identifier",
                            },
                            "type": "CallExpression",
                          },
                          "type": "ExpressionStatement",
                        },
                      ],
                      "directives": Array [],
                      "type": "BlockStatement",
                    },
                    "params": Array [],
                    "type": "ArrowFunctionExpression",
                  },
                  Object {
                    "elements": Array [
                      Object {
                        "name": "list",
                        "type": "Identifier",
                      },
                    ],
                    "type": "ArrayExpression",
                  },
                ],
                "callee": Object {
                  "name": "useEffect",
                  "type": "Identifier",
                },
                "type": "CallExpression",
              },
              "type": "ExpressionStatement",
            },
            Object {
              "argument": Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "closingElement": null,
                    "openingElement": Object {
                      "attributes": Array [
                        Object {
                          "name": Object {
                            "name": "onChange",
                            "type": "JSXIdentifier",
                          },
                          "type": "JSXAttribute",
                          "value": Object {
                            "expression": Object {
                              "name": "handleCurrentChange",
                              "type": "Identifier",
                            },
                            "type": "JSXExpressionContainer",
                          },
                        },
                        Object {
                          "name": Object {
                            "name": "value",
                            "type": "JSXIdentifier",
                          },
                          "type": "JSXAttribute",
                          "value": Object {
                            "expression": Object {
                              "name": "value",
                              "type": "Identifier",
                            },
                            "type": "JSXExpressionContainer",
                          },
                        },
                      ],
                      "name": Object {
                        "name": "input",
                        "type": "JSXIdentifier",
                      },
                      "selfClosing": true,
                      "type": "JSXOpeningElement",
                    },
                    "selfClosing": true,
                    "type": "JSXElement",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "expression": Object {
                          "name": "submitLabel",
                          "type": "Identifier",
                        },
                        "type": "JSXExpressionContainer",
                      },
                    ],
                    "closingElement": Object {
                      "name": Object {
                        "name": "button",
                        "type": "JSXIdentifier",
                      },
                      "type": "JSXClosingElement",
                    },
                    "openingElement": Object {
                      "attributes": Array [
                        Object {
                          "name": Object {
                            "name": "onClick",
                            "type": "JSXIdentifier",
                          },
                          "type": "JSXAttribute",
                          "value": Object {
                            "expression": Object {
                              "name": "handleClick",
                              "type": "Identifier",
                            },
                            "type": "JSXExpressionContainer",
                          },
                        },
                      ],
                      "name": Object {
                        "name": "button",
                        "type": "JSXIdentifier",
                      },
                      "selfClosing": false,
                      "type": "JSXOpeningElement",
                    },
                    "selfClosing": false,
                    "type": "JSXElement",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "expression": Object {
                          "arguments": Array [
                            Object {
                              "name": "TodoItem",
                              "type": "Identifier",
                            },
                            Object {
                              "name": "list",
                              "type": "Identifier",
                            },
                          ],
                          "callee": Object {
                            "name": "dynamicNode",
                            "type": "Identifier",
                          },
                          "type": "CallExpression",
                        },
                        "type": "JSXExpressionContainer",
                      },
                    ],
                    "closingElement": Object {
                      "name": Object {
                        "name": "ul",
                        "type": "JSXIdentifier",
                      },
                      "type": "JSXClosingElement",
                    },
                    "openingElement": Object {
                      "attributes": Array [],
                      "name": Object {
                        "name": "ul",
                        "type": "JSXIdentifier",
                      },
                      "selfClosing": false,
                      "type": "JSXOpeningElement",
                    },
                    "selfClosing": false,
                    "type": "JSXElement",
                  },
                ],
                "closingFragment": Object {
                  "type": "JSXClosingFragment",
                },
                "openingFragment": Object {
                  "type": "JSXOpeningFragment",
                },
                "type": "JSXFragment",
              },
              "type": "ReturnStatement",
            },
          ],
          "directives": Array [],
          "type": "BlockStatement",
        },
        "params": Array [],
        "type": "ArrowFunctionExpression",
      },
      "type": "VariableDeclarator",
    },
  ],
  "kind": "const",
  "type": "VariableDeclaration",
}
`;

exports[`statefulExample generates 1`] = `
"const App = () => {
  const valueKey = \\"value\\";
  const targetKey = \\"target\\";
  const listKey = \\"list\\";
  const currentKey = \\"current\\";
  const doneKey = \\"done\\";
  const falseKey = \\"false\\";
  const defaultCurrent = \\"\\";
  const textKey = \\"text\\";
  const defaultState = \\"[object Object]\\";
  const submitLabel = \\"Submit\\";
  const oneKey = 1;
  const zeroKey = 0;

  const getValueFromEvent = (state, event) => compose(objectOf(currentKey), get(valueKey), get(targetKey), nth(oneKey));

  const addTodo = (event, state) => compose(set(listKey, append(set(doneKey, falseKey, objectOf(textKey, get(currentKey, state))), get(listKey, state))), set(currentKey, defaultCurrent), constant(state));

  const [state, updateState] = useState(defaultState);
  const handleCurrentChange = useCallback(event => setState(getValueFromEvent(event, state)), [state]);
  const updateClick = useCallback(event => setState(addToDo(event, state)), [state]);
  const list = get(listKey, state);
  useEffect(() => {
    log(list, noop);
  }, [list]);
  return <><input onChange={handleCurrentChange} value={value} /><button onClick={handleClick}>{submitLabel}</button><ul>{dynamicNode(TodoItem, list)}</ul></>;
};"
`;

exports[`staticExample compiles 1`] = `
Object {
  "declarations": Array [
    Object {
      "id": Object {
        "name": "Button",
        "type": "Identifier",
      },
      "init": Object {
        "async": false,
        "body": Object {
          "body": Array [
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "hello",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "Hello",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "world",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "type": "StringLiteral",
                    "value": "World",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "declarations": Array [
                Object {
                  "id": Object {
                    "name": "label",
                    "type": "Identifier",
                  },
                  "init": Object {
                    "arguments": Array [
                      Object {
                        "name": "hello",
                        "type": "Identifier",
                      },
                      Object {
                        "name": "world",
                        "type": "Identifier",
                      },
                    ],
                    "callee": Object {
                      "name": "concat",
                      "type": "Identifier",
                    },
                    "type": "CallExpression",
                  },
                  "type": "VariableDeclarator",
                },
              ],
              "kind": "const",
              "type": "VariableDeclaration",
            },
            Object {
              "argument": Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "expression": Object {
                          "name": "label",
                          "type": "Identifier",
                        },
                        "type": "JSXExpressionContainer",
                      },
                    ],
                    "closingElement": Object {
                      "name": Object {
                        "name": "div",
                        "type": "JSXIdentifier",
                      },
                      "type": "JSXClosingElement",
                    },
                    "openingElement": Object {
                      "attributes": Array [],
                      "name": Object {
                        "name": "div",
                        "type": "JSXIdentifier",
                      },
                      "selfClosing": false,
                      "type": "JSXOpeningElement",
                    },
                    "selfClosing": false,
                    "type": "JSXElement",
                  },
                ],
                "closingFragment": Object {
                  "type": "JSXClosingFragment",
                },
                "openingFragment": Object {
                  "type": "JSXOpeningFragment",
                },
                "type": "JSXFragment",
              },
              "type": "ReturnStatement",
            },
          ],
          "directives": Array [],
          "type": "BlockStatement",
        },
        "params": Array [],
        "type": "ArrowFunctionExpression",
      },
      "type": "VariableDeclarator",
    },
  ],
  "kind": "const",
  "type": "VariableDeclaration",
}
`;

exports[`staticExample generates 1`] = `
"const Button = () => {
  const hello = \\"Hello\\";
  const world = \\"World\\";
  const label = concat(hello, world);
  return <><div>{label}</div></>;
};"
`;
